prov/efa/src/efa_user_info.c

```diff
@@ -269,34 +269,133 @@ int efa_user_info_get_dgram(uint32_t version, const char *node, const char *serv
 	return -FI_ENODATA;
 }
 
+#if HAVE_CUDA || HAVE_NEURON || HAVE_SYNAPSEAI
+/**
+ * @brief determine if EFA provider should claim support of FI_HMEM in info
+ * @param[in]	version		libfabric API version used by user
+ * @param[in]	hints		hints from user's call to fi_getinfo()
+ * @return	true, if EFA provider should claim support of FI_HMEM
+ * 		false, otherwise
+ */
+bool efa_user_info_should_support_hmem(int version, const struct fi_info *hints)
+{
+	/*
+	 * FI_HMEM is a primary capability. If user did not explicitly request
+	 * it, we should not cliam support of it.
+	 * only when user explicitly requeisted it.
+	 */
+	if (!hints || !(hints->caps & FI_HMEM))
+		return false;
+
+	/* Note that the default behavior of EFA provider is different between
+	 * libfabric API version.
+	 *
+	 * For libfabric API version 1.17 and earlier, EFA provider does not
+	 * support FI_HMEM unless GPUDirect RDMA is available.
+	 *
+	 * For libfabric API version 1.18 and later, EFA provider will claim support
+	 * of FI_HMEM as long as CUDA library is initialized and a CUDA device is
+	 * available. On an system without GPUDirect RDMA, the support of CUDA memory
+	 * is implemented by calling CUDA library. If user does not want EFA provider
+	 * to use CUDA library, the user can call use fi_setopt to set
+	 * FI_OPT_CUDA_API_PERMITTED to false. On an system without GPUDirect RDMA,
+	 * the call would fail.
```

Comment:
> On an system without GPUDirect RDMA, the call would fail.

IIUC this is trying to say on an(typo) system without gpu the user cannot set FI_OPT_CUDA_API_PERMITTED to false - is this a sensible behavior? Is there any problem to simply ignore this option in that case?

Reply:
>  without gpu the user cannot set FI_OPT_CUDA_API_PERMITTED to false - is this a sensible behavior?

They can try to set FI_OPT_CUDA_API_PERMITTED to false, the will get an error on system with out GPUDirect RDMA. Then they know that they cannot pass CUDA memory

>  Is there any problem to simply ignore this option in that case?

If we ignore. User will assume the provider is capable of supporting CUDA memory without calling cuda API. Then user will pass CUDA memory, which would cause hang.